File: ios/Classes/BlocklistManager.swift
import Foundation

/**
 * Manages a persistent blocklist of peers by sender UUID
 *
 * Blocked peers will not be able to connect (incoming or outgoing)
 * and their messages will be ignored.
 *
 * The blocklist is persisted to UserDefaults and survives app restarts.
 */
class BlocklistManager {
    private let tag = "BlocklistManager"

    // Thread-safe queue for blocklist operations
    private let queue = DispatchQueue(label: "com.ble_mesh.blocklist")

    // Set of blocked peer sender IDs
    private var blockedPeers: Set<String> = []

    // UserDefaults key
    private let blockedPeersKey = "ble_mesh_blocked_peers"

    init() {
        loadBlocklist()
    }

    /**
     * Block a peer by sender UUID
     *
     * - Parameter senderId: The sender UUID to block
     * - Returns: true if the peer was newly blocked, false if already blocked
     */
    func blockPeer(_ senderId: String) -> Bool {
        return queue.sync {
            if blockedPeers.insert(senderId).inserted {
                saveBlocklist()
                print("[\(tag)] Blocked peer: \(senderId) (total: \(blockedPeers.count))")
                return true
            }
            print("[\(tag)] Peer already blocked: \(senderId)")
            return false
        }
    }

    /**
     * Unblock a peer by sender UUID
     *
     * - Parameter senderId: The sender UUID to unblock
     * - Returns: true if the peer was unblocked, false if not blocked
     */
    func unblockPeer(_ senderId: String) -> Bool {
        return queue.sync {
            if blockedPeers.remove(senderId) != nil {
                saveBlocklist()
                print("[\(tag)] Unblocked peer: \(senderId) (remaining: \(blockedPeers.count))")
                return true
            }
            print("[\(tag)] Peer was not blocked: \(senderId)")
            return false
        }
    }

    /**
     * Check if a peer is blocked
     *
     * - Parameter senderId: The sender UUID to check
     * - Returns: true if the peer is blocked
     */
    func isBlocked(_ senderId: String) -> Bool {
        return queue.sync {
            return blockedPeers.contains(senderId)
        }
    }

    /**
     * Get all blocked peer IDs
     *
     * - Returns: Array of blocked sender UUIDs
     */
    func getBlockedPeers() -> [String] {
        return queue.sync {
            return Array(blockedPeers)
        }
    }

    /**
     * Get count of blocked peers
     *
     * - Returns: Number of blocked peers
     */
    func getBlockedPeerCount() -> Int {
        return queue.sync {
            return blockedPeers.count
        }
    }

    /**
     * Clear the entire blocklist
     */
    func clearBlocklist() {
        queue.sync {
            let count = blockedPeers.count
            blockedPeers.removeAll()
            saveBlocklist()
            print("[\(tag)] Cleared blocklist (\(count) peers removed)")
        }
    }

    /**
     * Load blocklist from persistent storage
     */
    private func loadBlocklist() {
        if let blockedArray = UserDefaults.standard.array(forKey: blockedPeersKey) as? [String] {
            blockedPeers = Set(blockedArray.filter { !$0.isEmpty })
            print("[\(tag)] Loaded \(blockedPeers.count) blocked peers from storage")
        } else {
            print("[\(tag)] No blocked peers found in storage")
        }
    }

    /**
     * Save blocklist to persistent storage
     */
    private func saveBlocklist() {
        let blockedArray = Array(blockedPeers)
        UserDefaults.standard.set(blockedArray, forKey: blockedPeersKey)
        UserDefaults.standard.synchronize()
        print("[\(tag)] Saved \(blockedPeers.count) blocked peers to storage")
    }

    /**
     * Clean up resources
     */
    func cleanup() {
        queue.sync {
            // Save before cleanup
            saveBlocklist()
            print("[\(tag)] Cleanup complete")
        }
    }
}

File: android/src/main/kotlin/com/ble_mesh/BlocklistManager.kt
package com.ble_mesh

import android.content.Context
import android.content.SharedPreferences
import android.util.Log
import java.util.concurrent.ConcurrentHashMap

/**
 * Manages a persistent blocklist of peers by sender UUID
 *
 * Blocked peers will not be able to connect (incoming or outgoing)
 * and their messages will be ignored.
 *
 * The blocklist is persisted to SharedPreferences and survives app restarts.
 */
class BlocklistManager(context: Context) {
    private val tag = "BlocklistManager"

    private val prefs: SharedPreferences = context.getSharedPreferences(
        PREFS_NAME,
        Context.MODE_PRIVATE
    )

    // Thread-safe set of blocked peer sender IDs
    private val blockedPeers = ConcurrentHashMap.newKeySet<String>()

    companion object {
        private const val PREFS_NAME = "ble_mesh_blocklist"
        private const val KEY_BLOCKED_PEERS = "blocked_peers"
    }

    init {
        loadBlocklist()
    }

    /**
     * Block a peer by sender UUID
     *
     * @param senderId The sender UUID to block
     * @return true if the peer was newly blocked, false if already blocked
     */
    fun blockPeer(senderId: String): Boolean {
        if (blockedPeers.add(senderId)) {
            saveBlocklist()
            Log.d(tag, "Blocked peer: $senderId (total: ${blockedPeers.size})")
            return true
        }
        Log.d(tag, "Peer already blocked: $senderId")
        return false
    }

    /**
     * Unblock a peer by sender UUID
     *
     * @param senderId The sender UUID to unblock
     * @return true if the peer was unblocked, false if not blocked
     */
    fun unblockPeer(senderId: String): Boolean {
        if (blockedPeers.remove(senderId)) {
            saveBlocklist()
            Log.d(tag, "Unblocked peer: $senderId (remaining: ${blockedPeers.size})")
            return true
        }
        Log.d(tag, "Peer was not blocked: $senderId")
        return false
    }

    /**
     * Check if a peer is blocked
     *
     * @param senderId The sender UUID to check
     * @return true if the peer is blocked
     */
    fun isBlocked(senderId: String): Boolean {
        return blockedPeers.contains(senderId)
    }

    /**
     * Get all blocked peer IDs
     *
     * @return List of blocked sender UUIDs
     */
    fun getBlockedPeers(): List<String> {
        return blockedPeers.toList()
    }

    /**
     * Get count of blocked peers
     *
     * @return Number of blocked peers
     */
    fun getBlockedPeerCount(): Int {
        return blockedPeers.size
    }

    /**
     * Clear the entire blocklist
     */
    fun clearBlocklist() {
        val count = blockedPeers.size
        blockedPeers.clear()
        saveBlocklist()
        Log.d(tag, "Cleared blocklist ($count peers removed)")
    }

    /**
     * Load blocklist from persistent storage
     */
    private fun loadBlocklist() {
        val blockedPeersString = prefs.getString(KEY_BLOCKED_PEERS, "") ?: ""
        if (blockedPeersString.isNotEmpty()) {
            val peerIds = blockedPeersString.split(",").filter { it.isNotEmpty() }
            blockedPeers.addAll(peerIds)
            Log.d(tag, "Loaded ${blockedPeers.size} blocked peers from storage")
        } else {
            Log.d(tag, "No blocked peers found in storage")
        }
    }

    /**
     * Save blocklist to persistent storage
     */
    private fun saveBlocklist() {
        val blockedPeersString = blockedPeers.joinToString(",")
        prefs.edit()
            .putString(KEY_BLOCKED_PEERS, blockedPeersString)
            .apply()
        Log.d(tag, "Saved ${blockedPeers.size} blocked peers to storage")
    }

    /**
     * Clean up resources
     */
    fun cleanup() {
        // Save before cleanup
        saveBlocklist()
        Log.d(tag, "Cleanup complete")
    }
}

